controllers/apps/cluster_plan_builder.go

```diff
@@ -110,10 +110,36 @@ func (c *clusterTransformContext) GetLogger() logr.Logger {
 	return c.Logger
 }
 
-func (c *clusterTransformContext) sharding(name string) bool {
-	// hack: use shardingComps to determine if the entity is sharding or component
-	_, ok := c.shardingComps[name]
-	return ok
+// sharding use to determine if the entity is sharding or component
+func (c *clusterTransformContext) sharding(name string) (bool, error) {
+	// sharding defined in cd topology
+	if len(c.Cluster.Spec.ClusterDef) > 0 && len(c.Cluster.Spec.Topology) > 0 {
```

Comment:
If either c.components or c.shardings have values, it indicates that the object has already been initialized, and we can directly determine it through c.shardingComps as a fast-path.

Reply:
> If either c.components or c.shardings have values, it indicates that the object has already been initialized, and we can directly determine it through c.shardingComps as a fast-path.

done